Enable multiple RTAI versions on a single system

/usr/bin/rtai-config becomes a stub script that helps select another
kernel-specific /usr/bin/rtai-config-<kver>.

https://github.com/ShabbyX/RTAI/issues/16

diff -ruN a/base/scripts/GNUmakefile.am b/base/scripts/GNUmakefile.am
--- a/base/scripts/GNUmakefile.am	2014-05-12 06:48:57.000000000 -0500
+++ b/base/scripts/GNUmakefile.am	2014-06-23 14:42:44.451696334 -0500
@@ -1,6 +1,8 @@
 install-exec-local:
 	$(mkinstalldirs) $(DESTDIR)$(bindir)
 	$(INSTALL_SCRIPT) rtai-load $(DESTDIR)$(bindir)
+	$(INSTALL_SCRIPT) rtai-config-kernel \
+		$(DESTDIR)$(bindir)/rtai-config-@RTAI_LINUX_UTSNAME@
 	$(INSTALL_SCRIPT) rtai-config $(DESTDIR)$(bindir)
 	$(INSTALL_SCRIPT) $(srcdir)/rtai-info $(DESTDIR)$(bindir)
 
diff -ruN a/base/scripts/rtai-config b/base/scripts/rtai-config
--- a/base/scripts/rtai-config	1969-12-31 18:00:00.000000000 -0600
+++ b/base/scripts/rtai-config	2014-06-23 22:20:11.586947316 -0500
@@ -0,0 +1,156 @@
+#!/bin/bash
+
+usage() {
+    test -n "$@" || echo "$@" >&2
+    {
+	echo "usage:  $0 OPTIONS ..."
+	echo
+	echo "This stub script locates and runs a kernel-specific version"
+	echo "of rtai-config."
+	echo
+	echo "Stub script options:"
+	echo "	-v KVER		Select rtai-config script for a kernel version"
+	echo "	-l		List available kernel choices"
+	echo "	-h | -?		This help"
+	echo ""
+	echo "kernel-specific options:"
+	echo "	--help"
+	echo "	--version"
+	echo "	--cc"
+	echo "	--cxx"
+	echo "	--arch"
+	echo "	--subarch"
+	echo "	--prefix"
+	echo "	--config"
+	echo "	--module-cflags"
+	echo "	--module-cxxflags"
+	echo "	--module-ext"
+	echo "	--lxrt-cflags"
+	echo "	--lxrt-ldflags"
+	echo "	--linux-dir"
+	echo "	--linux-version"
+	echo "	--module-dir"
+	echo "	--library-dir"
+	echo "	--comedi-dir"
+	echo "	--efltk-dir"
+	echo "	--posix-wrap"
+    } >&2
+    exit 1
+}
+
+select_rtai_config() {
+    # select_rtai_config -l lists all kernels an rtai-config is found
+    # for
+    if test "$1" = "-l"; then
+	local LIST=true
+	shift
+    else
+	local LIST=false
+    fi
+
+    # select_rtai_config -d searches for a kernel by its header
+    # directory location rather than its utsname
+    if test "$1" = "-d"; then
+	local FIND_BY_DIR=true
+	shift
+	# canonicalize header directory path from command line
+	local ARG="$(readlink -f $1)"
+    else
+	local FIND_BY_DIR=false
+	local ARG="$1"
+    fi
+
+    # loop through each rtai-config* executable in $PATH
+    for dir in ${PATH//:/ }; do
+	for c in $dir/rtai-config?*; do
+	    # Skip anything not executable
+	    test -x $c || continue
+
+	    if $FIND_BY_DIR; then
+		# Search for rtai-config with matching kernel header
+		# directory
+		TEST="$(readlink -f $($c --linux-dir 2>/dev/null))"
+	    else
+		# Search for rtai-config with matching kernel utsname
+		# or kernel version
+
+		# if rtai-config understands '--linux-utsname', use it
+		TEST="$($c --linux-utsname 2>/dev/null)"
+		# otherwise, use '--linux-version' option
+		test -n "$TEST" || TEST="$($c --linux-version)"
+	    fi
+
+	    if $LIST; then
+		# In list mode, list the choices
+		if test "$TEST" = "$RUNNING_KVER"; then
+		    # tell user this is the running kernel but keep
+		    # stdout readable by scripts
+		    echo -n "$TEST"
+		    echo -n "	** running kernel" >&2
+		    echo
+		else
+		    echo "$TEST"
+		fi
+
+	    else
+		# If the script output matches the given $ARG, print
+		# the path to rtai-config and return
+		if test "$TEST" = "$ARG"; then
+		    echo "$c"
+		    return
+		fi
+	    fi
+	done
+    done
+
+    # Finish up depending on mode
+    if $LIST; then
+	exit
+    else
+	echo "Error:  no rtai-config script found matching $ARG" >&2
+	exit 1
+    fi
+}
+
+# list of args for the kernel-specific rtai-config-$KVER script
+RTAI_CONFIG_ARGS=""
+
+# currently running kernel
+RUNNING_KVER="$(uname -r)"
+
+# be sure we don't accidentally call ourself
+test -z "$IN_RTAI_STUB" || { echo "Error:  ran ourself!" >&2 ; exit ; }
+export IN_RTAI_STUB=1
+
+# Usage if no command line args at all
+test -n "$1" || usage
+
+# Process command-line args; double-dash '--foo' arguments get passed
+# on to kernel rtai-config
+while getopts h?lv:d:-: ARG; do
+    case $ARG in
+	h | "?") usage ;;
+	v)  SELECT_ARG="$OPTARG" ;;
+	d)  SELECT_ARG="-d $OPTARG" ;;
+	l)  select_rtai_config -l ;;
+	-)  RTAI_CONFIG_ARGS="$RTAI_CONFIG_ARGS --$OPTARG" ;;
+    esac
+done
+
+# Usage if no args found to pass to kernel rtai-config
+test -n "$RTAI_CONFIG_ARGS" || usage
+
+# if no KVER selected, default to running kernel version and print
+# warning
+if test -z "$SELECT_ARG"; then
+    SELECT_ARG=$RUNNING_KVER
+    echo "Warning:  no '-v KVER' option specified" >&2
+    echo "Warning:  defaulting to running kernel version $SELECT_ARG" >&2
+    echo 'Warning:  (add "-v `uname -r`" to silence this warning)' >&2
+fi
+
+# select the kernel-specific rtai-config script by kernel version
+RTAI_CONFIG="$(select_rtai_config $SELECT_ARG)" || exit 1
+
+# run the selected rtai-config with supplied arguments
+$RTAI_CONFIG $RTAI_CONFIG_ARGS
diff -ruN a/base/scripts/rtai-config-kernel.in b/base/scripts/rtai-config-kernel.in
--- a/base/scripts/rtai-config-kernel.in	1969-12-31 18:00:00.000000000 -0600
+++ b/base/scripts/rtai-config-kernel.in	2014-06-23 14:55:59.783674630 -0500
@@ -0,0 +1,138 @@
+#! /bin/sh
+
+staging=${DESTDIR}
+prefix="@prefix@"
+exec_prefix="@exec_prefix@"
+libdir="@libdir@"
+datadir="@datadir@"
+pkgdatadir="${datadir}/@PACKAGE@"
+includedir="@includedir@"
+
+RTAI_VERSION="@PACKAGE_VERSION@"
+RTAI_PREFIX="${staging}${prefix}"
+RTAI_CC="@CC@"
+RTAI_CXX="@CXX@"
+RTAI_TARGET_ARCH="@RTAI_TARGET_ARCH@"
+RTAI_TARGET_SUBARCH="@RTAI_TARGET_SUBARCH@"
+RTAI_CONFIG="${staging}${pkgdatadir}/config-rtai-${RTAI_VERSION}"
+RTAI_LINUX_DIR="@RTAI_LINUX_DIR@"
+RTAI_LINUX_VERSION="@RTAI_LINUX_VERSION@"
+RTAI_LINUX_UTSNAME="@RTAI_LINUX_UTSNAME@"
+RTAI_KERNEL_CFLAGS="-I. -I${staging}${includedir} @RTAI_KMOD_APP_CFLAGS@ @RTAI_FP_CFLAGS@"
+RTAI_KERNEL_CXXFLAGS="-I. -I${staging}${includedir} @RTAI_KMOD_APP_CXXFLAGS@ @RTAI_FP_CFLAGS@"
+RTAI_MODULE_DIR="${staging}@RTAI_MODULE_DIR@"
+RTAI_LXRT_CFLAGS="-I. -I${staging}${includedir} @RTAI_USER_APP_CFLAGS@"
+RTAI_LXRT_LDFLAGS="-L${staging}${libdir} @RTAI_LXRT_LDADD@ -lpthread"
+RTAI_LIBRARY_DIR="${staging}${libdir}"
+COMEDI_DIR="@COMEDI_DIR@"
+EFLTK_DIR="@EFLTK_DIR@"
+RTAI_MODULE_EXT="@RTAI_MODULE_EXT@"
+RTAI_POSIX_WRAP="@RTAI_POSIX_WRAP@"
+
+unset prefix exec_prefix libdir datadir pkgdatadir includedir
+
+usage ()
+{
+cat <<EOF
+Usage rtai-config OPTIONS
+Options :
+        --help
+        --version
+        --cc
+        --cxx
+        --arch
+        --subarch
+        --prefix
+        --config
+        --module-cflags
+        --module-cxxflags
+        --module-ext
+        --lxrt-cflags
+        --lxrt-ldflags
+        --linux-dir
+        --linux-version
+        --linux-utsname
+        --module-dir
+        --library-dir
+        --comedi-dir
+        --efltk-dir
+        --posix-wrap
+EOF
+    exit $1
+}
+
+if test $# -eq 0; then
+    usage 1 1>&2
+fi
+
+while test $# -gt 0; do
+    case "$1" in
+        --version)
+            echo $RTAI_VERSION
+            ;;
+        --cc)
+            echo $RTAI_CC
+            ;;
+        --cxx)
+            echo $RTAI_CXX
+            ;;
+        --arch)
+            echo $RTAI_TARGET_ARCH
+            ;;
+        --subarch)
+            echo $RTAI_TARGET_SUBARCH
+            ;;
+        --prefix)
+            echo $RTAI_PREFIX
+            ;;
+        --datarootdir)
+            echo $RTAI_PREFIX
+            ;;
+        --config)
+            echo $RTAI_CONFIG
+            ;;
+        --mod*-cflags|--kernel-cflags)
+            echo $RTAI_KERNEL_CFLAGS
+            ;;
+        --mod*-ext)
+            echo $RTAI_MODULE_EXT
+            ;;
+        --mod*-cxxflags|--kernel-cxxflags)
+            echo $RTAI_KERNEL_CXXFLAGS
+            ;;
+        --lxrt-cflags|--user-cflags)
+            echo $RTAI_LXRT_CFLAGS
+            ;;
+        --lxrt-ldflags)
+            echo $RTAI_LXRT_LDFLAGS
+            ;;
+        --mod*-dir)
+            echo $RTAI_MODULE_DIR
+            ;;
+        --lib*-dir|--libdir|--user-libdir)
+            echo $RTAI_LIBRARY_DIR
+            ;;
+        --linux-dir|--linux)
+            echo $RTAI_LINUX_DIR
+            ;;
+        --linux-ver*)
+            echo $RTAI_LINUX_VERSION
+            ;;
+        --linux-utsname)
+            echo $RTAI_LINUX_UTSNAME
+            ;;
+        --comedi-dir)
+            echo $COMEDI_DIR
+            ;;
+        --efltk-dir)
+            echo $EFLTK_DIR
+            ;;
+        --posix-wrap)
+            echo $RTAI_POSIX_WRAP
+            ;;
+        *)
+         usage 1 1>&2
+         ;;
+    esac
+    shift
+done
diff -ruN a/base/scripts/rtai-config.in b/base/scripts/rtai-config.in
--- a/base/scripts/rtai-config.in	2014-05-12 06:48:57.000000000 -0500
+++ b/base/scripts/rtai-config.in	1969-12-31 18:00:00.000000000 -0600
@@ -1,133 +0,0 @@
-#! /bin/sh
-
-staging=${DESTDIR}
-prefix="@prefix@"
-exec_prefix="@exec_prefix@"
-libdir="@libdir@"
-datadir="@datadir@"
-pkgdatadir="${datadir}/@PACKAGE@"
-includedir="@includedir@"
-
-RTAI_VERSION="@PACKAGE_VERSION@"
-RTAI_PREFIX="${staging}${prefix}"
-RTAI_CC="@CC@"
-RTAI_CXX="@CXX@"
-RTAI_TARGET_ARCH="@RTAI_TARGET_ARCH@"
-RTAI_TARGET_SUBARCH="@RTAI_TARGET_SUBARCH@"
-RTAI_CONFIG="${staging}${pkgdatadir}/config-rtai-${RTAI_VERSION}"
-RTAI_LINUX_DIR="@RTAI_LINUX_DIR@"
-RTAI_LINUX_VERSION="@RTAI_LINUX_VERSION@"
-RTAI_KERNEL_CFLAGS="-I. -I${staging}${includedir} @RTAI_KMOD_APP_CFLAGS@ @RTAI_FP_CFLAGS@"
-RTAI_KERNEL_CXXFLAGS="-I. -I${staging}${includedir} @RTAI_KMOD_APP_CXXFLAGS@ @RTAI_FP_CFLAGS@"
-RTAI_MODULE_DIR="${staging}@RTAI_MODULE_DIR@"
-RTAI_LXRT_CFLAGS="-I. -I${staging}${includedir} @RTAI_USER_APP_CFLAGS@"
-RTAI_LXRT_LDFLAGS="-L${staging}${libdir} @RTAI_LXRT_LDADD@ -lpthread"
-RTAI_LIBRARY_DIR="${staging}${libdir}"
-COMEDI_DIR="@COMEDI_DIR@"
-EFLTK_DIR="@EFLTK_DIR@"
-RTAI_MODULE_EXT="@RTAI_MODULE_EXT@"
-RTAI_POSIX_WRAP="@RTAI_POSIX_WRAP@"
-
-unset prefix exec_prefix libdir datadir pkgdatadir includedir
-
-usage ()
-{
-cat <<EOF
-Usage rtai-config OPTIONS
-Options :
-        --help
-        --version
-        --cc
-        --cxx
-        --arch
-        --subarch
-        --prefix
-        --config
-        --module-cflags
-        --module-cxxflags
-        --module-ext
-        --lxrt-cflags
-        --lxrt-ldflags
-        --linux-dir
-        --linux-version
-        --module-dir
-        --library-dir
-        --comedi-dir
-        --efltk-dir
-        --posix-wrap
-EOF
-    exit $1
-}
-
-if test $# -eq 0; then
-    usage 1 1>&2
-fi
-
-while test $# -gt 0; do
-    case "$1" in
-        --version)
-            echo $RTAI_VERSION
-            ;;
-        --cc)
-            echo $RTAI_CC
-            ;;
-        --cxx)
-            echo $RTAI_CXX
-            ;;
-        --arch)
-            echo $RTAI_TARGET_ARCH
-            ;;
-        --subarch)
-            echo $RTAI_TARGET_SUBARCH
-            ;;
-        --prefix)
-            echo $RTAI_PREFIX
-            ;;
-        --datarootdir)
-            echo $RTAI_PREFIX
-            ;;
-        --config)
-            echo $RTAI_CONFIG
-            ;;
-        --mod*-cflags|--kernel-cflags)
-            echo $RTAI_KERNEL_CFLAGS
-            ;;
-        --mod*-ext)
-            echo $RTAI_MODULE_EXT
-            ;;
-        --mod*-cxxflags|--kernel-cxxflags)
-            echo $RTAI_KERNEL_CXXFLAGS
-            ;;
-        --lxrt-cflags|--user-cflags)
-            echo $RTAI_LXRT_CFLAGS
-            ;;
-        --lxrt-ldflags)
-            echo $RTAI_LXRT_LDFLAGS
-            ;;
-        --mod*-dir)
-            echo $RTAI_MODULE_DIR
-            ;;
-        --lib*-dir|--libdir|--user-libdir)
-            echo $RTAI_LIBRARY_DIR
-            ;;
-        --linux-dir|--linux)
-            echo $RTAI_LINUX_DIR
-            ;;
-        --linux-ver*)
-            echo $RTAI_LINUX_VERSION
-            ;;
-        --comedi-dir)
-            echo $COMEDI_DIR
-            ;;
-        --efltk-dir)
-            echo $EFLTK_DIR
-            ;;
-        --posix-wrap)
-            echo $RTAI_POSIX_WRAP
-            ;;
-        *)
-         usage 1 1>&2
-         ;;
-    esac
-    shift
-done
diff -ruN a/configure.ac b/configure.ac
--- a/configure.ac	2014-05-12 06:48:57.000000000 -0500
+++ b/configure.ac	2014-06-23 14:54:59.811676267 -0500
@@ -265,6 +265,26 @@
 fi
 
 dnl
+dnl check the utsname configured in the kernel headers
+dnl
+AC_MSG_CHECKING([linux utsname])
+if test -e $RTAI_LINUX_DIR/include/linux/utsrelease.h; then
+    VERSION_FILE=$RTAI_LINUX_DIR/include/linux/utsrelease.h
+elif test -e $RTAI_LINUX_DIR/include/generated/utsrelease.h; then
+    VERSION_FILE=$RTAI_LINUX_DIR/include/generated/utsrelease.h
+else
+    VERSION_FILE=$RTAI_LINUX_DIR/include/linux/version.h
+fi
+RTAI_LINUX_UTSNAME=`$CC -E -dM ${VERSION_FILE} | grep UTS | cut -d '"' -f 2`
+
+if test -z "$RTAI_LINUX_UTSNAME"; then
+    AC_MSG_RESULT([not found])
+else
+    AC_MSG_RESULT([$RTAI_LINUX_UTSNAME])
+fi
+
+
+dnl
 dnl define the name of the target arch (only matters for powerpc so far)
 dnl
 case $RTAI_TARGET_ARCH in
@@ -1928,6 +1948,7 @@
 AC_SUBST(RTAI_FP_CFLAGS)
 AC_SUBST(RTAI_LINUX_DIR)
 AC_SUBST(RTAI_LINUX_VERSION)
+AC_SUBST(RTAI_LINUX_UTSNAME)
 AC_SUBST(RTAI_MATH_LIBM)
 AC_SUBST(RTAI_MODULE_DIR)
 AC_SUBST(RTAI_MODULE_EXT)
@@ -1953,7 +1974,7 @@
 	base/sched/GNUmakefile \
 	base/sched/liblxrt/GNUmakefile \
 	base/scripts/GNUmakefile \
-	base/scripts/rtai-config \
+	base/scripts/rtai-config-kernel \
 	base/scripts/rtai-load \
 	base/arch/x86/GNUmakefile \
         base/arch/x86/hal/GNUmakefile \
